File: cpp/day3_part1_2016.cpp
#include <iostream>
#include <fstream>
#include <sstream>

bool isValidTriangle(int a, int b, int c) {
    return a + b > c && a + c > b && b + c > a;
}

int main() {
    std::ifstream file("input.txt");
    if (!file.is_open()) {
        std::cerr << "Error opening file" << std::endl;
        return 1;
    }

    std::string line;
    int validTriangles = 0;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        int a, b, c;
        if (!(iss >> a >> b >> c)) {
            std::cerr << "Invalid input format" << std::endl;
            continue;
        }

        if (isValidTriangle(a, b, c)) {
            validTriangles++;
        }
    }

    std::cout << validTriangles << std::endl;

    return 0;
}

File: haskell/day3_part1_2016.hs

main = do
    contents <- readFile "input.txt"
    let validTriangles = length $ filter isValidTriangle $ map parseInput $ lines contents
    print validTriangles

parseInput :: String -> (Int, Int, Int)
parseInput line = case map read $ words line of
    [a, b, c] -> (a, b, c)
    _ -> error "Invalid input format"

isValidTriangle :: (Int, Int, Int) -> Bool
isValidTriangle (a, b, c) = a + b > c && a + c > b && b + c > a
